### PR TITLE
feat: add mnemonic utility to check if a phrase is valid

### LINKED ISSUE #648: Mnemonic package should have a utility to check if a phrase is valid
Currently, when a user passes a seed phrase to the mnemonic component, the component doesn't check if the phrase is valid or not and accepts any sequence of existing words on the wordlist.

We should implement a validator utility to ensure the seed phrase is valid and return true or not.

Currently, this implementation is possible by just trying to create and expect an error, but we can implement a check using the wordlist.

```ts
export function isValidMnemonic(phrase: string): boolean {
  const words = phrase.split(' ');
  const mnemonic = new Mnemonic();
  return (
    words.filter((w) => mnemonic.wordlist.includes(w)).length === 12 // size of the mnemonic
  );
}
```

### PR DESCRIPTION
**Pull request by protocolwhisper**
Implemented IsValidMnemonic querying a wordlist with the binary search algorithm.
This Closes #648

### CODE REVIEW

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/mnemonic/src/wordlist.txt [hunk 1]
```diff
@@ -0,0 +1,2048 @@
+abandon
```
[the remark this answers is not in the document]
Comment (protocolwhisper) [reply packages/mnemonic/src/wordlist.txt:1]: Got it, I just update the import

[the remark this answers is not in the document]
Comment (luizstacio) [reply packages/mnemonic/src/wordlist.txt:1]: You can also remove the txt file

[the remark this answers is not in the document]
Comment (protocolwhisper) [reply packages/mnemonic/src/wordlist.txt]: Sure!!

File: packages/mnemonic/src/mnemonic.test.ts [hunk 2]
```diff
@@ -96,4 +96,17 @@ describe('Mnemonic', () => {
...
   });
+
+  test("Validate a incomplete Mnemonic phrase" , () => {
+    const mnemonicphrase = "Hello my name is Fuel"
+    const validation = Mnemonic.isValidMnemonic(mnemonicphrase);
+    expect(validation == false);
+  })
+
+  test("Validate a complete Mnemonic phrase" , () => {
+    const mnemonicphrase = "aware fringe excess tank asset quick suffer second cloth deliver file above"
```
Comment (luizstacio) [inline packages/mnemonic/src/mnemonic.test.ts:107]: Can you add tests for 12, 14, 16, 18, and 24 seed phrases?

**Review by luizstacio — CHANGES_REQUESTED** [review]

File: packages/mnemonic/src/mnemonic.test.ts (hunk 2 above)
Comment (protocolwhisper) [reply packages/mnemonic/src/mnemonic.test.ts:107]: Does >12 are valid in Fuel?

Comment (luizstacio) [reply packages/mnemonic/src/mnemonic.test.ts:107]: Yes, we accept any size of mnemonic. You can check on the method from Mnemonic.https://github.com/FuelLabs/fuels-ts/blob/d0eb1c732f63842b8d4801456054ec3b9ccdd020/packages/mnemonic/src/mnemonic.ts#L39

Comment (protocolwhisper) [reply packages/mnemonic/src/mnemonic.test.ts:113]: That's cool!! I added those and used that function

**Comment by protocolwhisper** [comment]
I just updated some issues with the Validate PR Title @luizstacio they were failing cause the name

File: packages/mnemonic/src/mnemonic.test.ts [hunk 3]
```diff
@@ -96,4 +96,36 @@ describe('Mnemonic', () => {
     const phrase = Mnemonic.generate(32, randomBytes(16));
     expect(phrase.split(' ').length).toBe(24);
   });
+
+
+  test("Validate a 12 complete Mnemonic phrase" , () => {
+    const mnemonicphrase = "aware fringe excess tank asset quick suffer second cloth deliver file above"
+    const validation = Mnemonic.isValidMnemonic(mnemonicphrase);
+    expect(validation === true);
```
Comment (arboleya) [inline packages/mnemonic/src/mnemonic.test.ts:104]:
```suggestion
    expect(validation).toEqual(true);
```

File: packages/mnemonic/src/mnemonic.test.ts [hunk 4]
```diff
@@ -96,4 +96,36 @@ describe('Mnemonic', () => {
...
+  })
+  test("Validate a 15 complete Mnemonic phrase" , () => {
+    const mnemonicphrase = "lucky buzz drink topic lunar proud warfare pass silver police chef liberty soap about process"
+    const validation = Mnemonic.isValidMnemonic(mnemonicphrase);
+    expect(validation === true);
```
Comment (arboleya) [inline packages/mnemonic/src/mnemonic.test.ts:109]:
```suggestion
    expect(validation).toEqual(true);
```

File: packages/mnemonic/src/mnemonic.test.ts [hunk 5]
```diff
@@ -96,4 +96,36 @@ describe('Mnemonic', () => {
...
+  test("Validate a 18 complete Mnemonic phrase" , () => {
+    const mnemonicphrase = "aware fringe excess tank asset quick suffer second cloth deliver file above"
+    const validation = Mnemonic.isValidMnemonic(mnemonicphrase);
+    expect(validation === true);
+  })
+
+  test("Validate a 21 complete Mnemonic phrase" , () => {
+    const mnemonicphrase = "movie fold other broccoli deliver bleak rookie record walnut coin decline blossom wheat thing silk group library absorb"
+    const validation = Mnemonic.isValidMnemonic(mnemonicphrase);
+    expect(validation === true);
```
Comment (arboleya) [inline packages/mnemonic/src/mnemonic.test.ts:121]:
```suggestion
    expect(validation).toEqual(true);
```

File: packages/mnemonic/src/mnemonic.test.ts [hunk 6]
```diff
@@ -96,4 +96,36 @@ describe('Mnemonic', () => {
...
+  })
+
+  test("Validate a 24 complete Mnemonic phrase" , () => {
+    const mnemonicphrase = "trick modify monster anger volcano thrive jealous lens warm program milk flavor bike torch fish eye aspect cable loan little bachelor town office sound"
+    const validation = Mnemonic.isValidMnemonic(mnemonicphrase);
+    expect(validation === true);
```
Comment (arboleya) [inline packages/mnemonic/src/mnemonic.test.ts:127]:
```suggestion
    expect(validation).toEqual(true);
```

**Review by arboleya — CHANGES_REQUESTED** [review]
Besides the comments and suggestions I left, the `lint` is failing:
 - https://github.com/FuelLabs/fuels-ts/actions/runs/3749796050/jobs/6368682873

**Comment by protocolwhisper** [comment]
> Besides the comments and suggestions I left, the `lint` is failing:
> 
> * https://github.com/FuelLabs/fuels-ts/actions/runs/3749796050/jobs/6368682873

Hello, I think it should be ok now with he lint errors

**Comment by protocolwhisper** [comment]
> Besides the comments and suggestions I left, the `lint` is failing:
> 
> * https://github.com/FuelLabs/fuels-ts/actions/runs/3749796050/jobs/6368682873

So I fix the prettier stuff, some imports, and the returns in the functions also the.toEqual in the test @arboleya :)

**Comment by arboleya** [comment]
@protocolwhisper Good one. Thank you for your contribution.

Let's wait for @luizstacio review.

**Review by arboleya — CHANGES_REQUESTED** [review]
@protocolwhisper I noticed the tests never assert for **_invalid_** mnemonics.

Could you please add some variations to cover these cases?

**Comment by protocolwhisper** [comment]
> @protocolwhisper I noticed the tests never assert for **_invalid_** mnemonics.
> 
> Could you please add some variations to cover these cases?

@arboleya Sure, I already pushed them lmk if I missed something

**Comment by protocolwhisper** [comment]
Fixed errors in the last two tests.
image

**Comment by protocolwhisper** [comment]
@luizstacio gm , am I missing smth in this PR?

**Review by arboleya — DISMISSED** [review]
Grammatically, I'd prefer naming the method `isMnemonicValid`. 

Other than that, looking good!

File: packages/mnemonic/src/mnemonic.ts [hunk 7]
```diff
@@ -130,6 +130,35 @@ class Mnemonic {
     return Mnemonic.masterKeysFromSeed(seed);
   }
 
+  static isValidMnemonic(phrase: string) {
```
Comment (Dhaiwat10) [inline packages/mnemonic/src/mnemonic.ts:133]:
```suggestion
  static isMnemonicValid(phrase: string) {
```

Like @arboleya said, I would suggest renaming the function

**Review by Dhaiwat10 — DISMISSED** [review]
LGTM! Just need to rename the function and we're good to go. Thanks :)

**Comment by Dhaiwat10** [comment]
I just pushed a commit to your branch that renames the function but it's not showing up here. Weird 🤔

**Comment by protocolwhisper** [comment]
@Dhaiwat10 I just committed it 🫶🏽

**Comment by Dhaiwat10** [comment]
@protocolwhisper you also need to update the test file! :)

**Comment by protocolwhisper** [comment]
> @protocolwhisper you also need to update the test file! :)

Gotcha!!  :+1:

**Comment by protocolwhisper** [comment]
@arboleya I need to wait for the issue to be solved?

**Review by arboleya — APPROVED** [review]

File: packages/mnemonic/src/mnemonic.ts [hunk 8]
```diff
@@ -130,6 +130,35 @@ class Mnemonic {
     return Mnemonic.masterKeysFromSeed(seed);
   }
 
+  static isMnemonicValid(phrase: string) {
+    // We can split the phrase and count the size of the array so we know if each array is valid or no
+
+    const mphrase = phrase.split(' ');
```
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/mnemonic/src/mnemonic.ts:136]: Nice catch!

[the remark this answers is not in the document]
Comment (arboleya) [reply packages/mnemonic/src/mnemonic.ts:136]: Will be addressed via:
 - https://github.com/FuelLabs/fuels-ts/issues/725

**Pull request merged**